man/vusu.1
.\" Copyright (C) 2019 VirtualSquare. Project Leader: Renzo Davoli
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License,
.\" as published by the Free Software Foundation, either version 2
.\" of the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
.\" MA 02110-1301 USA.
.\"
.\" Automatically generated by Pandoc 3.1.11
.\"
.TH "VUSU" "1" "November 2024" "VirtualSquare\-VUOS" "General Commands Manual"
.SH NAME
\f[CB]vusu\f[R] \[en] set the default networking stack
.SH SYNOPSIS
\f[CB]vusu\f[R] [\f[I]options\f[R]] [\-] [\f[I]user\f[R]
[*argment\&...]]
.SH DESCRIPTION
\f[CB]vusu\f[R] allows one to run commands with a substitute user and
group ID (in VUOS).
It requires a module loaded in VUOS able to redefine uid/gid,
e.g.\ \f[CB]unrealuidgid\f[R].
.PP
When called without arguments, \f[CB]vusu\f[R] defaults to running an
interactive shell as (virtual) \f[I]root\f[R].
.PP
For compatibility with \f[CB]su\f[R](1), \f[CB]vusu\f[R] defaults to not
change the current directory and to only set the environment variables
\f[CB]HOME\f[R] and \f[CB]SHELL\f[R] (plus \f[CB]USER\f[R] and
\f[CB]LOGNAME\f[R] if the target user is not root).
It is recommended to always use the \f[CB]\-\-login\f[R] option (instead
of its shortcut \-) to avoid side effects caused by mixing environments.
.SH OPTIONS
.TP
\f[CB]\-c\f[R] \f[I]command\f[R], \f[CB]\-\-command\f[R] \f[I]command\f[R]
Pass command to the shell with the \f[CB]\-c\f[R] option.
.TP
\f[CB]\-\f[R], \f[CB]\-l\f[R], \f[CB]\-\-login\f[R]
Start the shell as a login shell with an environment similar to a real
login:
.PD 0
.P
.PD
.TP
\f[CB]\f[R]
\- clears all the environment variables except \f[CB]TERM\f[R]
.PD 0
.P
.PD
.TP
\f[CB]\f[R]
\- initializes the environment variables \f[CB]HOME\f[R],
\f[CB]SHELL\f[R], \f[CB]USER\f[R], \f[CB]LOGNAME\f[R], and
\f[CB]PATH\f[R]
.PD 0
.P
.PD
.TP
\f[CB]\f[R]
\- changes to the target user\[cq]s home directory
.PD 0
.P
.PD
.TP
\f[CB]\f[R]
\- sets argv[0] of the shell to `\-' in order to make the shell a login
shell
.TP
\f[CB]\-m\f[R], \f[CB]\-p\f[R], \f[CB]\-\-preserve\-environment\f[R]
Preserve the entire environment, i.e.\ it does not set \f[CB]HOME\f[R],
\f[CB]SHELL\f[R], \f[CB]USER\f[R] nor \f[CB]LOGNAME\f[R].
This option is ignored if the option \f[CB]\-\-login\f[R] is specified.
.TP
\f[CB]\-s\f[R] \f[I]shell\f[R], \f[CB]\-\-shell\f[R] \f[I]shell\f[R]
Run the specified shell instead of the default.
The shell to run is selected according to the following
rules, in order:
.TP
\f[CB]\f[R]
\- the shell specified with \f[CB]\-\-shell\f[R]
.PD 0
.P
.PD
.TP
\f[CB]\f[R]
\- the shell specified in the environment variable \f[CB]SHELL\f[R], if
the \f[CB]\-\-preserve\-environment\f[R] option is used
.PD 0
.P
.PD
.TP
\f[CB]\f[R]
\- the shell listed in the passwd entry of the target user
.PD 0
.P
.PD
.TP
\f[CB]\f[R]
\- /bin/sh
.TP
\f[CB]\-h\f[R], \f[CB]\-\-help\f[R]
Display a short help message and exit.
.SH SEE ALSO
umvu(1), su(1)
.SH AUTHOR
VirtualSquare.
Project leader: Renzo Davoli.
(most of this man page is a derivative work from \f[CB]su\f[R](1) man
page)


cmd/vusu.c
/*
 *   VUOS: view OS project
 *   Copyright (C) 2018  Renzo Davoli <renzo@cs.unibo.it>
 *   VirtualSquare team.
 *
 *   This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include<stdio.h>
#include<stdlib.h>
#include<unistd.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <getopt.h>
#include <pwd.h>
#include <grp.h>
#include <string.h>
#include <libgen.h>
#include <vulib.h>

static char *short_options = "c:ls:mph";
static struct option long_options[] = {
	{"command", 1, 0, 'c'},
	{"login", 0, 0, 'l'},
	{"shell", 1, 0, 's'},
	{"preserve-environment", 0, 0, 'p'},
	{"help", 0, 0, 'h'},
	{0, 0, 0, 0}
};

char *command;
int login;
char *shell;
int change_environment = 1;
char *user;
int uid;
int gid;
char *arg0;
int ngroups;
gid_t *groups;

struct inheritenv {
	const char *tag;
	char *value;
} inheritenv[] = {
	{.tag = "TERM"},
	{.tag = "COLORTERM"},
	{.tag = "DISPLAY"},
	{.tag = "XAUTHORITY"},
	{NULL, NULL}
};

void usage(char *argv0)
{
	char *name=basename(argv0);
	fprintf(stderr,
			"Usage: %s [options] [LOGIN]\n"
			"\n"
			"Options:\n"
			"  -c, --command COMMAND         pass COMMAND to the invoked shell\n"
			"  -h, --help                    display this help message and exit\n"
			"  -, -l, --login                make the shell a login shell\n"
			"  -m, -p,\n"
			"  --preserve-environment        do not reset environment variables, and\n"
			"                                keep the same shell\n"
			"  -s, --shell SHELL             use SHELL instead of the default in passwd\n"
			"\n",
			name
			);
	exit(2);
}

char *getlogindefs(char *tag) {
	char *line = NULL;
	size_t linelen = 0;
	FILE *f=fopen("/etc/login.defs","r");
	char *retvalue = NULL;
	if (f) {
		while (retvalue == NULL && getline(&line, &linelen, f) > 0) {
			char *s = line;
			while (*s==' ' || *s=='\t') s++;
			if (*s=='#') continue;
			if (strncmp(tag,s,strlen(tag))!=0) continue;
			s+=strlen(tag);
			if (*s != ' ' && *s != '\t') continue;
			while (*s==' ' || *s=='\t') s++;
			s[strlen(s)-1]=0;
			retvalue = strdup(s);
		}
		fclose(f);
		if (line) free(line);
	}
	return retvalue;
}

void setpath(void)
{
	char *tag = (uid==0)?"ENV_SUPATH":"ENV_PATH";
	char *path = getlogindefs(tag);
	if (path) {
		char *cleanpath = path;
		if (strncmp("PATH=",cleanpath,5) == 0)
			cleanpath += 5;
		setenv("PATH",cleanpath,1);
		free(path);
	} else {
		if (uid)
			path="/bin:/usr/bin";
		else
			path="/sbin:/bin:/usr/sbin:/usr/bin";
		setenv("PATH",path,1);
	}
}

void loginenv(void) {
	int i;
	for (i = 0; inheritenv[i].tag != NULL; i++) {
		char *value = getenv(inheritenv[i].tag);
		if (value)
			inheritenv[i].value = strdup(value);
	}
	clearenv();
	for (i = 0; inheritenv[i].tag != NULL; i++) {
		if (inheritenv[i].value) {
			setenv(inheritenv[i].tag, inheritenv[i].value, 1);
			free(inheritenv[i].value);
		}
	}
}

int main(int argc, char *argv[])
{
	int c;
	struct passwd *pwd;

	if (vu_getinfo(NULL) < 0)
		execvp("su",argv);

	while (1) {
		int option_index = 0;
		c = getopt_long(argc, argv, short_options,
				long_options, &option_index);
		if (c == -1)
			break;
		switch (c) {
			case 'c': command=optarg;
								break;
			case 'l': login=1;
								break;
			case 's': shell=optarg;
								break;
			case 'm':
			case 'p': change_environment = 0;
								break;
			case 'h': usage(argv[0]);
								break;
			default:
								usage(argv[0]);
								break;
		}
	}

	if (argc > optind && strcmp(argv[optind],"-")==0) {
		login = 1;
		optind++;
	}

	if (argc > optind)
		user=argv[optind];

	if (argc > optind+1)
		usage(argv[0]);

	if (user == NULL)
		user = "root";

	pwd=getpwnam(user);
	if (pwd == NULL) {
		fprintf(stderr,"Unknown id: %s\n",user);
		exit(1);
	}
	uid=pwd->pw_uid;
	gid=pwd->pw_gid;
	if (login && change_environment)
		loginenv();

	if (shell == NULL) {
		if (change_environment)
			shell = pwd->pw_shell;
		else
			shell = getenv("SHELL");
	}
	if (shell == NULL)
		shell="/bin/sh";
	unsetenv("IFS");
	if (change_environment) {
		setenv("USER",user,1);
		setenv("LOGNAME",user,1);
		setenv("HOME",pwd->pw_dir,1);
		setenv("SHELL",shell,1);
	}
	if (login == 0 || asprintf(&arg0,"-%s",shell) < 0)
		arg0=shell;
	getgrouplist(user,gid,NULL,&ngroups);
	groups=malloc(ngroups * sizeof (gid_t));
	if (groups == NULL)
		ngroups=0;
	else
		getgrouplist(user,gid,groups,&ngroups);

	if (setresuid(uid,uid,uid) < 0 ||
			setresgid(gid,gid,gid) < 0)
		perror(argv[0]);
	else {
		setgroups(ngroups,groups);
		setpath();
		if (login) {
			int defhome = 0;
			char *defhomeopt = getlogindefs("DEFAULT_HOME");
			if (defhomeopt) {
				if (strcmp(defhomeopt, "yes") == 0)
					defhome = 1;
				free(defhomeopt);
			}
			if (chdir(pwd->pw_dir) < 0) {
				if (defhome == 0 || chdir("/") < 0) {
					perror(argv[0]);
					exit(1);
				}
			}
		}
		if (command)
			execl(shell,arg0,"-c",command,(char *)0);
		else
			execl(shell,arg0,(char *)0);
		perror(arg0);
	}
	exit(1);
	return 0;
}
